### ISSUE TITLE
type-aware lint loses bun:test resolution and cascades into no-unsafe-call false positives

### ISSUE THREAD

**Issue by dallenpyrah** [oxc-project/tsgolint#852]
## Summary

`oxlint` type-aware linting appears to lose `bun:test` module resolution in a real workspace configuration, which then cascades into large numbers of false-positive `typescript/no-unsafe-call` and `typescript/no-unsafe-member-access` diagnostics in Bun test files.

In the affected workspace:

- `tsc` can type-check Bun test files
- `oxlint --type-check` reports `TS2307: Cannot find module 'bun:test' or its corresponding type declarations`
- after that unresolved import, `describe(...)`, `it(...)`, `expect(...)`, and matcher chains become `error`-typed and trigger `no-unsafe-*` findings

## Versions

- `oxlint`: `1.57.0`
- `oxlint-tsgolint`: `0.17.4`
- `typescript`: `5.9.3`
- `bun`: `1.3.9`
- `bun-types`: `1.3.9`

## What I expected

`bun:test` should resolve during type-aware linting the same way it does for TypeScript, so Bun test APIs should not become `error`-typed.

## What actually happens

Running this command against a Bun test file:

```bash
bunx oxlint --type-check -c .context/tsgolint-buntest-repro/oxlint.json --tsconfig tsconfig.oxlint.json packages/ai/tests/tools/chart/schemas.test.ts
```

produces:

```text
x typescript(TS2307): Cannot find module 'bun:test' or its corresponding type declarations.
, -[packages/ai/tests/tools/chart/schemas.test.ts:1:38]
1 | import { describe, expect, it } from "bun:test";
  :                                      ^^^^^^^^^^
```

and then cascades into many false positives like:

```text
x typescript-eslint(no-unsafe-call): Unsafe call of a(n) `error` type typed value.
x typescript-eslint(no-unsafe-member-access): Unsafe member access .toBe on an `error` typed value.
```

on `describe(...)`, `it(...)`, `expect(...)`, `.toBe(...)`, `.toThrow(...)`, etc.

## Additional evidence

This standalone Bun test file type-checks correctly with TypeScript:

```ts
import { describe, expect, it } from "bun:test";

describe("bun:test typing", () => {
  it("keeps expect typed", () => {
    expect(true).toBe(true);
  });
});
```

and:

```bash
bunx tsc -p .context/tsgolint-buntest-repro/tsconfig.json --noEmit
```

passes with no errors.

So the problem does not appear to be Bun's own type definitions.

## Notes

- I was not able to reduce this all the way down to a tiny standalone `oxlint` repro yet.
- A tiny standalone repro with just `bun:test` and a minimal tsconfig passes.
- The failure does reproduce consistently on the real workspace tsconfig and test files.
- In the affected repo, adding a **test-only override** for `typescript/no-unsafe-call` and `typescript/no-unsafe-member-access` removes a large set of Bun-test false positives, but that is only a workaround.

If helpful, I can try to reduce this further, but I wanted to report the concrete `TS2307` + `bun:test` resolution failure first because that appears to be the root of the downstream noise.

**Comment by jakebailey** [comment]
You should be testing TS 6.0. It sounds like you don't have `types` set in your `tsconfig` to include `bun`.

**Comment by dallenpyrah** [comment]
I tested that hypothesis directly.

Standalone repro results with the modern Bun setup:
- TypeScript: `6.0.0-dev.20260324`
- `@types/bun` installed
- `types: ["bun"]`

That setup passes for the small `bun:test` repro under both `tsc` and `oxlint`.

I also tested the real workspace case with a repo-derived tsconfig switched over to `types: ["bun"]`, and `oxlint --type-check` still reports:

```
TS2307: Cannot find module "bun:test" or its corresponding type declarations.
```

So I agree the issue report should account for TS 6 / `types: ["bun"]`, but this does not appear to be just a missing-`types` config problem. The failure still reproduces on the real workspace-shaped config/file combination.

If helpful, I can keep reducing toward a smaller repro from there.

**Comment by dallenpyrah** [comment]
Additional narrowing:

I reduced the trigger further. The import graph is not required.

A brand-new file containing only:

```ts
import { describe, expect, it } from "bun:test";

describe("bun:test resolution", () => {
  it("keeps bun test symbols typed", () => {
    expect(true).toBe(true);
  });
});
```

still reproduces the failure under `oxlint --type-check` **when the file lives inside a workspace package subtree** (for example `packages/ai/...`).

Key comparison with the same minimal oxlint config and same tsconfig shape:
- repo root file: `tsc` passes, `oxlint` passes
- nested non-package dir file: `tsc` passes, `oxlint` passes
- `packages/ai/...` file: `tsc` passes, `oxlint` reports `TS2307` for `bun:test` and then cascades into `no-unsafe-call` / `no-unsafe-member-access`

So the narrowest trigger I have so far is: **files under a workspace package subtree**. This now looks less like Bun typing in general and more like package-subtree / workspace-scoped type resolution in the type-aware program.

**Comment by dallenpyrah** [comment]
I now have a much smaller repro.

Clean external workspace setup:
- root `package.json` with `workspaces: ["packages/*"]`
- root `tsconfig.json` using TypeScript 6-style Bun config (`types: ["bun"]`)
- root `oxlint.json` enabling only `typescript/no-unsafe-call` and `typescript/no-unsafe-member-access` with `options.typeAware: true`
- `root.test.ts` importing `bun:test`
- `packages/plain/package.test.ts` importing `bun:test`

Baseline: both files pass under `oxlint --type-check`.

Trigger: add an **empty** file at `packages/plain/tsconfig.json`:

```json
{}
```

Then rerun the exact same command for `packages/plain/package.test.ts`, and it starts failing with:

```
TS2307: Cannot find module "bun:test" or its corresponding type declarations.
```

while `root.test.ts` still passes, and `tsc -p tsconfig.json --noEmit` still passes for the full workspace.

So the smallest trigger I have now is:
- workspace package subdir file
- plus a nearer package-local `tsconfig.json` (even empty)

That strongly suggests type-aware resolution is switching to / discovering the nearer tsconfig for package-subtree files despite the outer root tsconfig being the one passed on the command line.

**Comment by camc314** [comment]
@dallenpyrah As per the docs, the `--tsconfig` flag isn't respected for using type checking https://oxc.rs/docs/guide/usage/linter/cli.html (I know this is a bit confusing - unfortunately due to backwards compatibility, we need to keep it.

TS Config resolutions works the exact same way as when your using typescript inside your editor. In your case it does this by:
1. walk up directories until it finds a tsconfig.json
2. check the file we're linting is inside that tsconfig (not blocked by includes/excludes)
3. else continue walking
(there's other complexities around project references, but the TLDR is we do the same thing as in your editor.

In your case, when you add `packages/plain/tsconfig.json`, it means that tsconfig is being used to lint that file - as noted, in TypeScript 6.0, types defaults to an empty array - to fix this yuo need to add `types: ["bun"]` to the tsconfig.json.

Given this, I would be suprised if you were not seeing errors inside your editor when the test file is open.

That said, the easiest way to fix this is to add `types: ["bun"]` to that nested tsconfig - there are other solutions, but this is the simplest.

This appears to be working as intended - please try the above suggestion and let me know if it gets you working

**Comment by camc314** [comment]
@dallenpyrah closing as not planned.

Please comment or make a new issue if you run into other issue, or reach out on discord if you need help configuring your project.

Thanks
